packages/devkit/src/generators/typescript/config.ts

```diff
@@ -0,0 +1,71 @@
+import type { Tree } from 'nx/src/generators/tree';
+import { writeJson } from 'nx/src/generators/utils/json';
+import type { JsonSerializeOptions } from 'nx/src/utils/json';
+import { dirname } from 'path';
+
+let ts: typeof import('typescript');
+
+/**
+ * Reads the provided tsconfig json file.
+ *
+ * @param tree File system tree.
+ * @param tsConfigPath Path to the tsconfig json file.
+ * @returns The parsed tsconfig json file.
+ */
+export function readTsConfigJson<T extends object = any>(
+  tree: Tree,
+  tsConfigPath: string
+): T {
+  if (!ts) {
+    ts = require('typescript');
+  }
+
+  const { config } = ts.readConfigFile(tsConfigPath, (path) =>
+    tree.read(path, 'utf-8')
+  );
+
+  return config;
+}
+
+/**
+ * Reads the full TS configuration from the provided tsconfig json file path
+ * by loading the base configuration if set.
+ *
+ * @param tree File system tree.
+ * @param tsConfigPath Path to the tsconfig json file.
+ * @returns The parsed TS configuration.
+ */
+export function readTsConfiguration<T extends object = any>(
+  tree: Tree,
+  tsConfigPath: string
+): T {
+  const config = readTsConfigJson(tree, tsConfigPath);
+
+  return ts.parseJsonConfigFileContent(
+    config,
+    ts.sys,
+    dirname(tsConfigPath)
+  ) as T;
+}
+
+/**
+ * Updates a tsconfig json file.
+ *
+ * @param tree File system tree.
+ * @param tsConfigPath Path to the tsconfig json file.
+ * @param updater Function that maps the current value of the tsconfig json file
+ * to a new value to be written to the file.
+ * @param options Options to be use when serializing the JSON.
+ */
+export function updateTsConfigJson<
+  T extends object = any,
+  U extends object = T
+>(
+  tree: Tree,
+  tsConfigPath: string,
+  updater: (value: T) => U,
+  options?: JsonSerializeOptions
+): void {
+  const updatedValue = updater(readTsConfigJson(tree, tsConfigPath));
+  writeJson(tree, tsConfigPath, updatedValue, options);
```

Comment:
This means we would strip trailing comma and comments from the file when we update it right? Is that OK?

There have been a few issues re: jsonc recently, and this was my main hiccup when thinking about it. See #10393

Reply:
You're right. I think is tricky to come up with a good approach that takes everything into account. With the `jsonc-parser` we can achieve writing a json and keeping the comments and trailing commas, but the DX is not the greatest:

```js
const jsonText = tree.read(filePath, 'utf-8');
const edits = modify(jsonText, ['foo'], 'baz', { allowTrailingComma: true });
const updatedJsonText = applyEdits(tree.read('index.json', 'utf-8'), edits);
tree.write(filePath, updatedJsonText);
```

A generic method with that sort of API would need to return an array of "modifications" instead of the json (rough code, things can be improved, but it should give an idea):

```js
// helper

import {
  applyEdits,
  Edit,
  ModificationOptions,
  modify,
  parse,
  ParseOptions,
  printParseErrorCode,
} from 'jsonc-parser';

export interface JsonModification {
  path: string[];
  value?: any;
}

export function updateJson<T extends object = any>(
  tree: Tree,
  path: string,
  updater: (value: T) => JsonModification[],
  options: ParseOptions & ModificationOptions = {}
): void {
  const fileText = tree.read(path, 'utf-8');
  const errors = [];
  const json = parse(fileText, errors, options);
  if (errors.length) {
    const { error, offset } = errors[0];
    throw new Error(
      `${printParseErrorCode(error)} in JSON at position ${offset}`
    );
  }

  const edits: Edit[] = [];
  const modifications = updater(json);
  modifications.forEach(({ path, value }) => {
    edits.push(...modify(fileText, path, value, options));
  });
  const updatedJsonText = applyEdits(fileText, edits);
  tree.write(path, updatedJsonText);
}

// usage

function doSomething(tree: Tree) {
  updateJson(
    tree,
    'a.json',
    (json) => {
      const modifications: JsonModification[] = [
        // overwrites/set a value regardless of whether it exists or not
        { path: ['a'], value: true },
        // removes a value
        { path: ['b'], value: undefined },
      ];

      // conditionally set something
      if (json.c?.d) {
        modifications.push({ path: ['c', 'd'], value: 'e' });
      }

      return modifications;
    },
    { allowTrailingComma: true }
  );
}
```